internal/server/polling/server.go

```diff
@@ -105,66 +113,161 @@ func (s *Server) poll(ctx context.Context, resource types.NamespacedName, secret
 }
 
-func (s *Server) reconcile(ctx context.Context, original *infrav1.Terraform, source *sourcev1.GitRepository, prs []provider.PullRequest) error {
-	// List Terraform objects, created by the branch planner.
-	tfList, err := s.listTerraformObjects(ctx, original.Namespace, map[string]string{
-		bpconfig.LabelKey: bpconfig.LabelValue,
-	})
-	if err != nil {
-		return fmt.Errorf("failed to list Terraform objects: %w", err)
-	}
+func (s *Server) reconcile(ctx context.Context, original *infrav1.Terraform, source *sourcev1.GitRepository, prs []provider.PullRequest, gitProvider provider.Provider) error {
+	s.log.Info("starting reconciliation ...")
 
-	// Create and delete objects based on the diff
-	// between PRs and Terraform objects.
+	// Create a map of pull requests, with the PR number as the key and the PR itself as the value.
 	prMap := map[string]provider.PullRequest{}
 	for _, pr := range prs {
-		prMap[fmt.Sprintf("%d", pr.Number)] = pr
+		prId := fmt.Sprintf("%d", pr.Number)
+		prMap[prId] = pr
+		s.log.Info("mapping PR", "PR number", pr.Number)
 
-		if err = s.reconcileTerraform(ctx, original, source, pr.HeadBranch, fmt.Sprintf("%d", pr.Number), s.branchPollingInterval); err != nil {
+		// Reconcile the Terraform objects related to each PR.
+		// If an error occurs, log it and continue with the next PR.
+		if err := s.reconcileTerraform(ctx, original, source, pr.HeadBranch, prId, s.branchPollingInterval); err != nil {
 			s.log.Error(err, "failed to reconcile Terraform object")
+		} else {
+			s.log.Info("successfully reconciled Terraform object for PR", "PR number", pr.Number)
 		}
 	}
 
-	for _, branchTf := range tfList {
-		prLabel := branchTf.Labels[bpconfig.LabelPRIDKey]
+	// List the Terraform planner objects in the namespace of the original object
+	s.log.Info("listing Terraform objects...")
+	tfPlannerObjects, err := s.listTerraformObjects(ctx, original.Namespace, map[string]string{
+		bpconfig.LabelKey: bpconfig.LabelValue,
+	})
 
-		if _, ok := prMap[prLabel]; !ok {
-			if err = s.deleteTerraform(ctx, branchTf); err != nil {
+	// If an error occurs, wrap it with context information and return it.
+	if err != nil {
+		return fmt.Errorf("failed to list Terraform objects: %w", err)
+	}
+
+	s.log.Info("iterating over Terraform planner objects...")
+	// For each Terraform object created by the branch planner,
+	// check if there's a corresponding open PR. If not, delete the Terraform object.
+	for _, tfPlannerObject := range tfPlannerObjects {
+		prId := tfPlannerObject.Labels[bpconfig.LabelPRIDKey]
+		pr, exist := prMap[prId]
+		// If the PR does not exist or has been closed, delete the related Terraform object.
+		// If an error occurs, log it.
+		if !exist || pr.Closed {
+			s.log.Info("the PR either does not exist or has been closed, deleting corresponding Terraform object...", "PR ID", prId)
+			if err = s.deleteTerraform(ctx, tfPlannerObject); err != nil {
 				s.log.Error(err, "failed to delete Terraform object")
```

Comment:
Same as above, which terraform? Maybe it's in the `err` error message, maybe it's not, we should not rely on it.